src/main/java/org/openstreetmap/atlas/utilities/identifiers/EntityIdentifierGenerator.java

```diff
@@ -0,0 +1,147 @@
+package org.openstreetmap.atlas.utilities.identifiers;
+
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.openstreetmap.atlas.exception.CoreException;
+import org.openstreetmap.atlas.geography.atlas.complete.CompleteEdge;
+import org.openstreetmap.atlas.geography.atlas.complete.CompleteEntity;
+import org.openstreetmap.atlas.geography.atlas.complete.CompleteNode;
+import org.openstreetmap.atlas.geography.atlas.complete.CompleteRelation;
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.utilities.collections.StringList;
+
+/**
+ * Generate unique 64 bit (Java long) identifiers for {@link CompleteEntity}s. The identifiers are
+ * generated using a hash of the entity's properties, including its geometry and tags. While the
+ * identifiers are advertised as unique, 64 bits may not be enough to prevent collisions when used
+ * at world-scale.
+ * 
+ * @author lcram
+ */
+public class EntityIdentifierGenerator
+{
+    private static final long HIGHEST_ATLAS_ID = 900000000999999L;
+    private static final long LOWEST_ATLAS_ID = -900000000999999L;
+
+    /**
+     * Generate a 64 bit hash for a given {@link CompleteEntity}. The entity must contain enough
+     * information for it to be created from scratch.
+     * 
+     * @param entity
+     *            the entity
+     * @return the hash
+     */
+    public long generate(final CompleteEntity<?> entity)
+    {
+        final String entityString = getPropertyString(entity)
+                + getTypeSpecificPropertyString(entity);
+        UUID entityHash = UUID.nameUUIDFromBytes(entityString.getBytes());
+
+        long shortHash = entityHash.getMostSignificantBits();
+        while (!isHashSafeToUse(shortHash))
```

Comment:
Want to add a safeguard here, to protect the unlikely event of a (close to?) infinite loop?